### PR TITLE
Refactor to target .NET Standard 1.1

### PR DESCRIPTION
**Pull request by 6A**
Huge refactor with two main goals:
- Target .NET Standard 1.1, and thus Windows, macOS and Linux.
- Improve code to adhere more to the C# coding guidelines.

This breaks many things, sorry.

### CODE REVIEW

**Comment by aquynh** [comment]
@chaplin89, can you please ack?

**Comment by 6A** [comment]
Additionally, I'd like to publish it as a NuGet package, in which case I also need to specify the authors and copyrights ; what should I put for those? I suppose both @chaplin89 and I should be in the authors if my push is accepted, but you might want me to add other contributors.

**Comment by aquynh** [comment]
yes, it is between you & @chaplin89 to decide what to put into the author list. same for the license of the NuGet package.

cheers.

File: bindings/csharp/Keystone.Net/Engine.cs [hunk 1]
```diff
@@ -3,150 +3,160 @@
 using System.Runtime.InteropServices;
 using System.Threading;
 
-namespace KeystoneNET
+namespace Keystone
```
Comment (chaplin89) [inline bindings/csharp/Keystone.Net/Engine.cs]: Your renaming (this and the others) makes sense to me but you're broking everything (as you said).
Anyhow, for me it's ok as long it's ok also for @aquynh that maybe has a clearer visibility on how many people are using this.

Comment (6A) [reply bindings/csharp/Keystone.Net/Engine.cs]: Here is why I made these changes:
- [There shouldn't be uppercase parts in names](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions), and I think `Keystone` looks better than `KeystoneNet`. It conflicts with the [Keystone Framework](https://www.nuget.org/packages/Keystone), but to me it doesn't matter, since their [toplevel namespace doesn't contain many members](http://www.keystoneframework.com/docs/index.html#frlrfKeystone.html) (thus, no conflict in code).
- It's a little redundant to specify `KeystoneEngine` as opposed to `Engine`, for example. Plus `Engine` conveys the purpose better.
- The library has never been published on NuGet, which means that everyone using it has gotten it from GH. As such, major changes such as this one are expected, and should hapen sooner rather than later.
- Some enums and enum members haven't been renamed because they're not used directly in wrapped, managed types. Maybe I should change this?

File: bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature [hunk 2]
```diff
@@ -1,20 +0,0 @@
-﻿Feature: Compilation
```
[the remark this answers is not in the document]
Comment (6A) [reply bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature]: The testing framework you used was really nice, but unfortunately is [not yet compatible with .NET Core / Standard](http://specflow.org/documentation/.NET-Support/).

File: bindings/csharp/Keystone.Net/NativeInterop.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+using System.Runtime.InteropServices;
+
+namespace Keystone
+{
+    /// <summary>
+    ///   Imported symbols for interop with keystone.dll.
+    /// </summary>
+    internal class NativeInterop
+    {
+        // This shouldn't be needed, even on Windows
```
[the remark this answers is not in the document]
Comment (6A) [reply bindings/csharp/Keystone.Net/NativeInterop.cs]: My computer did not need it on Windows / AnyCPU. Furthermore, I'm pretty sure it will make it crash on non-Windows platforms. Finally, [LLVMSharp](https://github.com/Microsoft/LLVMSharp) does not need any such initializer, so I doubt it's really useful now.

File: bindings/csharp/Keystone.Net/Engine.cs (hunk 1 above)
Comment (chaplin89) [reply bindings/csharp/Keystone.Net/Engine.cs]: No need to explain, for me it was already ok. Just see what @aquynh think.

File: bindings/csharp/Keystone.Net/NativeInterop.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chaplin89) [reply bindings/csharp/Keystone.Net/NativeInterop.cs]: Since it's "AnyCPU", since you're running it on an x86_64 platform, chances are your app is ngen'd into 64 bit machine code so if, for instance, you have a 64 bit version of your keystone.dll in your output dir, everything works correctly.

Say you've built your app with AnyCPU and your app is executed on a 32 bit platform, the application is ngen'd into 32 bit machine code, so this will no longer work because you have the 64 bit version of your app.

File: bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature (hunk 2 above)
[the remark this answers is not in the document]
Comment (chaplin89) [reply bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature]: I'm not very expert about code/standard and the like but on their website I read only about ".net core".

Specflow is composed of an extension of VS that only serve to generate the *.feature's code behind (this should not pose any compatibility issues) and internally it uses NUnit (and also this should not pose any compatibility issues). Also, I guess you can target .NET "something" in your test app and .net standard in everything else? Anyhow, I should try.

[the remark this answers is not in the document]
Comment (6A) [reply bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature]: .NET Core replaces the Framework and runs everywhere instead of just Windows; .NET Standard replaces Portable Class Libraries. Thus SpecFlow currently only targets Windows, and requires VS (except Express) or msbuild. We could use .NET Framework for the tests since it supports .NET Standard, but this would mean that tests are only available on Windows.

Finally, the generated code uses NUnit **and** SpecFlow (see [here](https://github.com/keystone-engine/keystone/pull/344/files/3eba118b711b807774d7be3a59b992ce02581387#diff-3e0003d485808a442aceea5f2196d7ceL5)).

File: bindings/csharp/Keystone.Net/NativeInterop.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (6A) [reply bindings/csharp/Keystone.Net/NativeInterop.cs]: I unfortunately cannot test it, but that seems really strange to me, especially since I don't know any other binding library that requires this. If you want, we could add an opt-in static method that does this, in case it is needed?

File: bindings/csharp/Keystone.Net/Engine.cs (hunk 1 above)
Comment (6A) [reply bindings/csharp/Keystone.Net/Engine.cs]: [He gave his opinion there](https://github.com/keystone-engine/keystone/pull/344#issuecomment-377777761), so the decision is yours to make.

**Comment by aquynh** [comment]
Yes, i will merge upon Marco's approval.

**Comment by chaplin89** [comment]
Ok for me, then

File: bindings/csharp/Keystone.Net/NativeInterop.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chaplin89) [reply bindings/csharp/Keystone.Net/NativeInterop.cs]: If other libraries doesn't use something like that, they leave to whoever build the final package the burden to choose the right "native" dll  (x86/x64), and that's a problem with AnyCPU code because you don't know if it will be x86 or x64 until runtime. But that's fine I guess. 
If the choice is between portability and supporting this edge case, portability wins.

File: bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature (hunk 2 above)
[the remark this answers is not in the document]
Comment (chaplin89) [reply bindings/csharp/KeystoneNET/KeystoneNET.Tests/Compilation.feature]: Ok then, altough I'm still pretty sure you can have the "test" assembly compiled with a given version of the framework, and hence supporting specflow, and all the rest compiled in .net standard.
But that's fine anyhow, as long as there are unit tests.

**Comment by 6A** [comment]
If that interests you guys, I can make a small markdown file that explains all the changes regarding to the API (with examples) for the few people that use the existing bindings.

**Comment by chaplin89** [comment]
That's ok, although I think is pretty clear how to use the lib. 
As you like for the authors file, I don't have any preferences.

**Comment by aquynh** [comment]
so it is all clear, that i can merge this now?

**Comment by 6A** [comment]
Personally I'm good, I'll change the Authors field when I publish the package on NuGet; I'll probably put his username there

**Comment by aquynh** [comment]
merged, thanks!
